### PR TITLE
Make AtomicRef behave for like std::atomic

### PR DESCRIPTION
**Pull request by MrBurmark**
Make RAJA::AtomicRef behave more like std::atomic.

Add `load` and `operator T` to load values. Should we have an `operator T` at all or should we force the use of `.load()`?
Add `store` alongside existing `operator=` to store values.
Added `compare_exchange_strong` and `compare_exchange_weak` alongside existing `CAS`.
Change `op=` operators to return the new stored value `old op val`.
Added `fetch_op` methods that return the previously stored value `old`.

Added tests for new operators.

### CODE REVIEW

**Comment by trws** [comment]
I'd say we want to match the to-be-std [atomic_ref](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0019r8.html) which offers these.  Some things that std::atomic has, and does, don't make sense for ref but I see no reason not to offer these.

File: test/unit/test-atomic.cpp [hunk 1]
```diff
@@ -100,91 +100,524 @@ void testAtomicFunctionPol()
...
+  PreIncCountOp(T* count, RAJA::RangeSegment seg)
+    : counter(count), min((T)0), max((T)seg.size()-(T)1), final((T)seg.size())
+  { count[0] = (T)0; }
+  RAJA_HOST_DEVICE
+  T operator()(RAJA::Index_type RAJA_UNUSED_ARG(i)) const {
+    return (++counter) - (T)1;
+  }
+  RAJA::atomic::AtomicRef<T, AtomicPolicy> counter;
+  T min, max, final;
+};
```
Comment (rhornung67) [inline test/unit/test-atomic.cpp]: @MrBurmark minor nit. Please add a blank line between each of the struct definitions in this file so it's easier to read.

File: test/unit/test-atomic.cpp [hunk 2]
```diff
@@ -100,91 +100,524 @@ void testAtomicFunctionPol()
...
+template < typename T >
+RAJA_INLINE
+RAJA_HOST_DEVICE
+typename std::enable_if<sizeof(T) == 1, T>::type np2m1(T val)
+{
+  val |= val >> 1  ;
+  val |= val >> 2  ;
+  val |= val >> 4  ;
+  return val;
+}
```
Comment (rhornung67) [inline test/unit/test-atomic.cpp]: Please add a blank line between these things too.

**Comment by MrBurmark** [comment]
@trws @ajkunen I fiddled with atomicCAS to get it to work with floating point, could you take a look?

**Comment by trws** [comment]
The logic looks ok, but now that you mention it why the rename to builtin_atomic_cas?

**Comment by MrBurmark** [comment]
Before that that method was the outward facing atomicCAS method for builtin.
I changed the name of the implementation to differentiate it from the outward facing atomicCAS method for builtin.
I also wanted to check that we are all ok with using std::enable_if in the implementation. I don't think that I've seen it used much in the code.

**Pull request merged**